eth/Delegation_Voting_EIP-712.md

```diff
@@ -1,233 +1,285 @@
+
+
+# Cast Vote By Signature in a Web3 Site
+# 在Web3中通过签名投票
+
+Just like in the Delegate By Signature example, a web page can be made for users to create a vote signature. If a user wishes to give away their “ballot” to a third party, that user must sign separate “for” and “against” transactions, and the third-party may choose which transaction to publish.
+
+和委托一样，投票也可以创建一个网页来完成。用户在委托投票权利的时候，“赞同”和“反对”的委托需要分开转账，第三方可以选择最终发布哪一个委托。
+
+The structured data object for creating vote signatures is slightly different from a delegate signature. In this case, the user needs a “Ballot” definition.
+这个结构化数据和上面的委托签名稍许不同，这里用户需要定义“Ballot”
+
+```
+const createVoteBySigMessage = (govAddress, proposalId, support, chainId = 1) => {
+  const types = {
+    EIP712Domain: [
+      { name: 'name', type: 'string' },
+      { name: 'chainId', type: 'uint256' },
+      { name: 'verifyingContract', type: 'address' },
+    ],
+    Ballot: [
+      { name: 'proposalId', type: 'uint256' },
+      { name: 'support', type: 'bool' }
+    ]
+  };
+  const primaryType = 'Ballot';
+  const domain = { name: 'Compound Governor Alpha', chainId, verifyingContract: govAddress };
+  support = !!support;
+  const message = { proposalId, support };
+  return JSON.stringify({ types, primaryType, domain, message });
+};
+```
+
+Clicking “Create Vote Signatures” will prompt the user to sign a “for” and an “against” transaction. Similar to the **delegateBySig** method, votes can be cast by passing 2 parameters to the **castVoteBySig** method.
+
+点击“创建投票签名（Create Vote Signatures）”之后，会弹出让用户选择签名“赞成”和“反对”转账。与**delegateBySig**函数一致，通过传递2个参数来调用**castVoteBySig**函数。
+
+1. The unique ID of the Compound governance proposal (auto-incrementing integer).
+1. Compound管理提案唯一编号（自动增加）
+2. A boolean value of the user’s support of the proposal (true or false).
+2. 布尔值来表示赞成还是反对提案
+
+Additionally, the user must pass the signature, which must be broken up into 3 parameters, known as **v**, **r**, and **s** (implemented in the previous section).
+另外用户还要传入分成3部分的签名，通常分为**v**, **r**和**s**
+
+The [full web page code example](https://github.com/compound-developers/compound-governance-examples/tree/master/signature-examples) for casting a vote with an EIP-712 signature is available in the [governance examples GitHub repository](https://github.com/compound-developers/compound-governance-examples).
+
+按照EIP-712规范投票[完整页面代码](https://github.com/compound-developers/compound-governance-examples/tree/master/signature-examples)可以在[compound管理代码仓库](https://github.com/compound-developers/compound-governance-examples)查看。
+
+# Batch and Publish Signatures with a Script or a Smart Contract
+# 通过脚本或者智能合约打包和发布签名
+
+Once a user has collected signed transactions for Compound governance, they need to publish those transactions to the Ethereum blockchain. The code for implementing this could be useful for exchanges and wallet apps that manage many private keys for many different users.

```

Comment:
Compound governance  \   transactions

Reply:
The code for implementing this could be useful for exchanges and wallet apps that manage many private keys for many different users.  理解有偏差， 代码不是管理私钥的。 管理多个用户的私钥 是用来修饰 exchanges及 wallet